### PR TITLE
Make all backends controlled by features (using a very ugly workaround)

### LINKED ISSUE #3514: Add a default feature for wgpu that enables Vulkan & GLES on Windows & Linux
We currently only enable backends based on the target platform with the only exception being GLES which is always available.

We should expose a new feature flag (enabled by default) that covers the GLES and DX11 backends so that users not interested in those backends can avoid compiling them in their binaries.

Feature name ideas:
- `legacy`
- `compat` (mapping to a future WebGPU Compat; might not be the best choice since we support even lower limits/features than what WebGPU Compat is aiming for; i.e. GLES 3.1+ and DX11 FL10+)

See background for this: https://github.com/gfx-rs/wgpu/issues/1221#issuecomment-1430067962

### PR DESCRIPTION
**Pull request by Wumpf**
**Connections**
* Fixes https://github.com/gfx-rs/wgpu/issues/3514
* Very ugly workaround for https://github.com/rust-lang/cargo/issues/1197

**Description**
Feature flag changes:
* new `vulkan` feature for enabling Vulkan on Windows/Android/Linux (part of default feature set)
* new `gles` feature for  enabling Vulkan on Windows/Android/Linux (part of default feature set, debatable if that's a good idea)
* fix `webgl` enabling GL on all platforms

This is achieved by introducing three proxy crates. Each depends on wgpu-core under a different platform condition and will "forward" all feature flags to it.

Reminder why this horrible construct is the only way:
This is necessary because it's impossible today to drag in a dependency conditionally BOTH on a platform & a feature flag.
It is possible to do all these things in code, so we can forward features all the way to wgpu-hal and not have it compile e.g. it's vulkan backend on Mac iff `vulkan-portability` is off. **But** without this workaround it's impossible to avoid building vulkan associated dependencies anyways.
For more details see
* https://github.com/rust-lang/cargo/issues/1197

Building with `default-features=false` will now not even drag in wgpu-core! `cargo tree -p wgpu --edges normal` for simple application depending on wgpu this way:
```
wgpu v24.0.0 (D:\dev\wgpu\wgpu)
├── arrayvec v0.7.6
├── bitflags v2.8.0
├── document-features v0.2.10 (proc-macro)
│   └── litrs v0.4.1
├── log v0.4.25
├── parking_lot v0.12.3
│   ├── lock_api v0.4.12
│   │   └── scopeguard v1.2.0
│   └── parking_lot_core v0.9.10
│       ├── cfg-if v1.0.0
│       ├── smallvec v1.13.2
│       └── windows-targets v0.52.6
│           └── windows_x86_64_msvc v0.52.6
├── profiling v1.0.16
├── raw-window-handle v0.6.2
├── smallvec v1.13.2
├── static_assertions v1.1.0
└── wgpu-types v24.0.0 (D:\dev\wgpu\wgpu-types)
    ├── bitflags v2.8.0
    └── log v0.4.25
```

Similarly, only depending on `dx12` (on Windows) doesn't include  `ash` or `glow` (the former is a slow building dependency!):
```
wgpu v24.0.0 (D:\dev\wgpu\wgpu)
├── arrayvec v0.7.6
├── bitflags v2.8.0
├── document-features v0.2.10 (proc-macro)
│   └── litrs v0.4.1
├── log v0.4.25
├── parking_lot v0.12.3
│   ├── lock_api v0.4.12
│   │   └── scopeguard v1.2.0
│   └── parking_lot_core v0.9.10
│       ├── cfg-if v1.0.0
│       ├── smallvec v1.13.2
│       └── windows-targets v0.52.6
│           └── windows_x86_64_msvc v0.52.6
├── profiling v1.0.16
├── raw-window-handle v0.6.2
├── smallvec v1.13.2
├── static_assertions v1.1.0
├── wgpu-core v24.0.0 (D:\dev\wgpu\wgpu-core)
│   ├── arrayvec v0.7.6
│   ├── bit-vec v0.8.0
│   ├── bitflags v2.8.0
│   ├── document-features v0.2.10 (proc-macro) (*)
│   ├── hashbrown v0.15.2
│   │   └── foldhash v0.1.4
│   ├── indexmap v2.7.0
│   │   ├── equivalent v1.0.1
│   │   └── hashbrown v0.15.2 (*)
│   ├── log v0.4.25
│   ├── naga v24.0.0 (D:\dev\wgpu\naga)
│   │   ├── arrayvec v0.7.6
│   │   ├── bit-set v0.8.0
│   │   │   └── bit-vec v0.8.0
│   │   ├── bitflags v2.8.0
│   │   ├── codespan-reporting v0.11.1
│   │   │   ├── termcolor v1.4.1
│   │   │   │   └── winapi-util v0.1.9
│   │   │   │       └── windows-sys v0.59.0
│   │   │   │           └── windows-targets v0.52.6 (*)
│   │   │   └── unicode-width v0.1.14
│   │   ├── indexmap v2.7.0 (*)
│   │   ├── log v0.4.25
│   │   ├── rustc-hash v1.1.0
│   │   ├── strum v0.26.3
│   │   │   └── strum_macros v0.26.4 (proc-macro)
│   │   │       ├── heck v0.5.0
│   │   │       ├── proc-macro2 v1.0.93
│   │   │       │   └── unicode-ident v1.0.14
│   │   │       ├── quote v1.0.38
│   │   │       │   └── proc-macro2 v1.0.93 (*)
│   │   │       ├── rustversion v1.0.19 (proc-macro)
│   │   │       └── syn v2.0.96
│   │   │           ├── proc-macro2 v1.0.93 (*)
│   │   │           ├── quote v1.0.38 (*)
│   │   │           └── unicode-ident v1.0.14
│   │   ├── termcolor v1.4.1 (*)
│   │   └── thiserror v2.0.11
│   │       └── thiserror-impl v2.0.11 (proc-macro)
│   │           ├── proc-macro2 v1.0.93 (*)
│   │           ├── quote v1.0.38 (*)
│   │           └── syn v2.0.96 (*)
│   ├── once_cell v1.20.2
│   ├── parking_lot v0.12.3 (*)
│   ├── profiling v1.0.16
│   ├── raw-window-handle v0.6.2
│   ├── rustc-hash v1.1.0
│   ├── smallvec v1.13.2
│   ├── thiserror v2.0.11 (*)
│   ├── wgpu-hal v24.0.0 (D:\dev\wgpu\wgpu-hal)
│   │   ├── arrayvec v0.7.6
│   │   ├── bit-set v0.8.0 (*)
│   │   ├── bitflags v2.8.0
│   │   ├── gpu-allocator v0.27.0
│   │   │   ├── log v0.4.25
│   │   │   ├── presser v0.3.1
│   │   │   ├── thiserror v1.0.69
│   │   │   │   └── thiserror-impl v1.0.69 (proc-macro)
│   │   │   │       ├── proc-macro2 v1.0.93 (*)
│   │   │   │       ├── quote v1.0.38 (*)
│   │   │   │       └── syn v2.0.96 (*)
│   │   │   └── windows v0.58.0
│   │   │       ├── windows-core v0.58.0
│   │   │       │   ├── windows-implement v0.58.0 (proc-macro)
│   │   │       │   │   ├── proc-macro2 v1.0.93 (*)
│   │   │       │   │   ├── quote v1.0.38 (*)
│   │   │       │   │   └── syn v2.0.96 (*)
│   │   │       │   ├── windows-interface v0.58.0 (proc-macro)
│   │   │       │   │   ├── proc-macro2 v1.0.93 (*)
│   │   │       │   │   ├── quote v1.0.38 (*)
│   │   │       │   │   └── syn v2.0.96 (*)
│   │   │       │   ├── windows-result v0.2.0
│   │   │       │   │   └── windows-targets v0.52.6 (*)
│   │   │       │   ├── windows-strings v0.1.0
│   │   │       │   │   ├── windows-result v0.2.0 (*)
│   │   │       │   │   └── windows-targets v0.52.6 (*)
│   │   │       │   └── windows-targets v0.52.6 (*)
│   │   │       └── windows-targets v0.52.6 (*)
│   │   ├── hashbrown v0.15.2 (*)
│   │   ├── libloading v0.8.6
│   │   │   └── windows-targets v0.52.6 (*)
│   │   ├── log v0.4.25
│   │   ├── naga v24.0.0 (D:\dev\wgpu\naga) (*)
│   │   ├── once_cell v1.20.2
│   │   ├── ordered-float v4.6.0
│   │   │   └── num-traits v0.2.19
│   │   ├── parking_lot v0.12.3 (*)
│   │   ├── profiling v1.0.16
│   │   ├── range-alloc v0.1.4
│   │   ├── raw-window-handle v0.6.2
│   │   ├── rustc-hash v1.1.0
│   │   ├── thiserror v2.0.11 (*)
│   │   ├── wgpu-types v24.0.0 (D:\dev\wgpu\wgpu-types)
│   │   │   ├── bitflags v2.8.0
│   │   │   └── log v0.4.25
│   │   ├── windows v0.58.0 (*)
│   │   └── windows-core v0.58.0 (*)
│   └── wgpu-types v24.0.0 (D:\dev\wgpu\wgpu-types) (*)
└── wgpu-types v24.0.0 (D:\dev\wgpu\wgpu-types) (*)
```

**Testing**

* separate test app with various config options
   * TODO: try some more

**Checklist**

- [x] Run `cargo fmt`.
- [x] Run `taplo format`.
- [x] Run `cargo clippy`. If applicable, add:
  - [x] `--target wasm32-unknown-unknown`
  - [ ] `--target wasm32-unknown-emscripten`
- [ ] Run `cargo xtask test` to run tests.
- [ ] Add change to `CHANGELOG.md`. See simple instructions inside file.

--- 
Draft todo:
* [x] do we even want this abomination
* [ ] fix lotsa unused warnings
* [ ] document proxy crates
* [x] should GL really be in the default feature set
* [ ] test this on Mac & Linux by hand
* [ ] check on emscripten and other corner cases
* [ ] surely ci fails ;-)
* [ ] write changelog

### CODE REVIEW

**Review by cwfitzgerald — COMMENTED** [review]
I think this is broadly fine, I will have some comments on specifics once this is ready for review, but we definitely shouldn't be scared of crates. Might be worthwhile looking into cargo-release or release-plz or something if the count grows.

**Comment by sagudev** [comment]
> Building with default-features=false will now not even drag in wgpu-core! cargo tree -p wgpu --edges normal for simple application depending on wgpu this way

This is important for https://github.com/gfx-rs/wgpu/issues/6330, I even impl it as https://github.com/gfx-rs/wgpu/commit/faafd37f999c209591e5e703c8f249b502980653.

File: Cargo.toml [hunk 1]
```diff
@@ -20,6 +20,9 @@ members = [
     "wgpu-macros",
     "wgpu-types",
     "wgpu",
+    "wgpu-core-target-feature-conditionals/arch-wasm32",
+    "wgpu-core-target-feature-conditionals/native-non-apple",
+    "wgpu-core-target-feature-conditionals/vendor-apple",
```
Comment (brody2consult) [inline Cargo.toml]: Could we consider shorter names for these, perhaps with slight reordering?

```suggestion
    "wgpu-core-target-features/arch-wasm32",
    "wgpu-core-target-features/native-non-apple",
    "wgpu-core-target-features/vendor-apple",
    "wgpu",
```

File: wgpu/src/api/instance.rs [hunk 2]
```diff
@@ -60,45 +60,23 @@ impl Instance {
...
-            if cfg!(target_os = "macos") && cfg!(feature = "angle") {
-                backends = backends.union(Backends::GL);
-            }
-        } else {
-            if cfg!(webgpu) {
-                backends = backends.union(Backends::BROWSER_WEBGPU);
-            }
-            if cfg!(webgl) {
-                backends = backends.union(Backends::GL);
-            }
```
Comment (brody2consult) [inline wgpu/src/api/instance.rs:99]: <3 always love for removing large blocks of ugly code :)

**Review by brody2consult — COMMENTED** [review]
quick comments from an outsider

File: Cargo.toml (hunk 1 above)
Comment (Wumpf) [reply Cargo.toml]: yeah those names are definitely too crazy right now

Comment (cwfitzgerald) [reply Cargo.toml]: I also have thoughts on the naming scheme that I'll fully think through once this is out of draft

**Comment by jimblandy** [comment]
@Wumpf My impassioned request is that the explanation in the top of this PR make it into the code as a comment. I agree that this a good PR, but people are going to be utterly WTH when they see this.

**Comment by jimblandy** [comment]
Also: spectacular

**Comment by cwfitzgerald** [comment]
Sorry, kinda stepped on your toes a bit with #6980 - lmk if you need help managing the migration

**Comment by Wumpf** [comment]
hit a bit of a snag: any feature that is supposed to be forwarded to wgpu-core needs to be forwarded now to all four (yeah it's four by now) proxy crates. Doable, but _extra_ annoying

The problem is that if I put `wgpu-core = {workspace=true, optional=true}` but then always address it with `some-feature = ["wgpu-core?/some-feature"]` it has no effect since after all nothing actually enables `wgpu-core`.
